docxcompose/properties.py

```diff
@@ -198,96 +199,197 @@ def set_properties(self, properties):
+
+
+class FieldBase(object):
+
+    fieldname_search_expr = re.compile(
+            r'DOCPROPERTY +"{0,1}([^\\]*?)"{0,1} +\\\* MERGEFORMAT',
+            flags=re.UNICODE)
+
+    def __init__(self, field_node):
+        self.node = field_node
+        self.name = self._get_fieldname()
+
+    def _format_value(self, value):
         if isinstance(value, bool):
-            value = u'Y' if value else u'N'
+            return u'Y' if value else u'N'
         elif isinstance(value, datetime):
-            value = value.strftime('%x')
+            return value.strftime('%x')
         else:
-            value = text_type(value)
-
-        # Simple field
-        sfield = xpath(
-            self.doc.element.body,
-            u'.//w:fldSimple[contains(@w:instr, \'DOCPROPERTY "{}"\')]'.format(name))
-        if sfield:
-            text = xpath(sfield[0], './/w:t')
+            return text_type(value)
+
+    def update(self, value):
+        raise NotImplementedError()
+
+    def remove_field_keep_value(self):
+        raise NotImplementedError()
+
+    def _get_fieldname_string(self):
+        raise NotImplementedError()
+
+    def _get_fieldname(self):
+        match = self.fieldname_search_expr.search(self._get_fieldname_string())
+        if match is None:
+            return None
+        return match.groups()[0]
+
+
+class RunNode(object):
+
+    def __init__(self, node, is_begin=False, is_end=False,
+                 is_separate=False, is_after_separate=False):
+        self.node = node
+        self.is_begin = is_begin
+        self.is_end = is_end
+        self.is_separate = is_separate
+        self.is_after_separate = is_after_separate
+
+
+class SimpleField(FieldBase):
+
+    attr_name = "{{{}}}instr".format(NS["w"])
+
+    def _get_fieldname_string(self):
+        return self.node.attrib[self.attr_name]
+
+    def update(self, value):
+        text = xpath(self.node, './/w:t')
+        if text:
+            text[0].text = self._format_value(value)
+
+    def remove_field_keep_value(self):
+        parent = self.node.getparent()
+        index = list(parent).index(self.node)
+        w_r = deepcopy(self.node[0])
+        parent.remove(self.node)
+        parent.insert(index, w_r)
+
+
+class ComplexField(FieldBase):
+
+    def __init__(self, field_node):
+        super(ComplexField, self).__init__(field_node)
+        # run and paragraph containing the field
+        self.w_r = self.node.getparent()
+        self.w_p = self.w_r.getparent()
+
+    def _get_fieldname_string(self):
+        return self.node.text
+
+    def _get_field_bound_indexes(self):
+        """Searches for the begin and end tags surrounding the complex
+        field as well as the corresponding position of the separate tag.
+        Returns their position in the parent node"""
+        fieldname_index = self.w_p.index(self.w_r)
+        begins = [self.w_p.index(el.getparent()) for el in xpath(self.w_p, u'.//w:r/w:fldChar[@w:fldCharType="begin"]')]
```

Comment:
i think you could simplify searching for the sibling nodes by using the `preceding-sibling` or `following-sibling` xpath axis, starting with `self.w_r`.